doschema/jsvalidation.py

```diff
@@ -0,0 +1,232 @@
+        self.fields_types_dict = {}
+        self.uri = None
+        self.resolver_factory = resolver_factory or jsonschema.RefResolver
+
+    def start(self, schema, uri):
+        """Start processing a given list of schemas.
+
+        :param schema: Name of schema that is currently processed.
+        :param uri: Name of schema that is currently processed.
+        """
+        self.uri = uri
+        self.resolver = self.resolver_factory(base_uri=uri, referrer=schema)
+
+        self.traverse(schema, ())
+
+        self.uri = None
+        self.resolver = None
+
+    def traverse(self, curr_schema, curr_field, check_dependencies=True):
+        """Go through the schema and retrieve schema's particular fields.
+
+        :param curr_schema: Schema or subschema that is currently processed.
+        :param curr_field: Tuple with path to currently processed field.
+        :param check_dependencies:
+        """
+        self.check_special(curr_schema, curr_field)
+
+        while '$ref' in curr_schema:
+            path = curr_schema.get('$ref')
+            curr_schema = self.resolver.resolve(path)[1]
+
+        if curr_schema.get('type') == 'object':
+            if 'properties' in curr_schema:
+                props = curr_schema['properties']
+
+                for prop in props:
+                    if isinstance(props[prop], dict):
+                        while '$ref' in props[prop]:
+                            path = props[prop].get('$ref')
+                            props[prop] = self.resolver.resolve(path)[1]
+
+                    prop_type = props[prop].get('type')
+
+                    self.process_type_field(
+                        prop_type,
+                        curr_field + (prop, )
+                    )
+
+                    if prop_type == 'object':
+                        self.traverse(
+                            props[prop],
+                            curr_field + (prop, )
+                        )
+                    elif prop_type == 'array':
+                        self.process_array(
+                            props[prop]['items'],
+                            curr_field
+                        )
+
+            if check_dependencies and 'dependencies' in curr_schema:
+                self.traverse_dependencies(curr_schema, curr_field)
+
+        else:
+            raise ValueError('Curr_schema type not an object')
+
+    def process_array(self, field_value, curr_field):
+        """Process each JSON object in the array according to ignore_index.
+
+        :param field_value: Type of the field.
+        :param curr_field: Tuple with path to currently processed field.
+        """
+        field_path = curr_field
+        for elem in field_value:
+            if self.ignore_index:
+                new_field_path = field_path
+            else:
+                new_field_path = field_path + (field_value.index(elem), )
+
+            self.traverse(elem, new_field_path)
```

Comment:
The array values can have other types than `object`. Thus `traverse` can't be called directly on them.